edgehog-device-runtime-docker/src/store/mod.rs

```diff
@@ -0,0 +1,202 @@
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+// SPDX-License-Identifier: Apache-2.0
+
+//! Persistent stores of the request issued by Astarte and resources created.
+
+use std::borrow::Cow;
+use std::io::{self, Cursor, SeekFrom};
+use std::path::Path;
+
+use serde::{Deserialize, Serialize};
+use tokio::fs::File;
+use tokio::io::{AsyncBufReadExt, AsyncSeekExt, AsyncWriteExt, BufReader, BufWriter};
+
+use crate::service::node::Nodes;
+use crate::service::Node;
+
+/// Error returned by the [`StateStore`].
+#[non_exhaustive]
+#[derive(Debug, thiserror::Error, displaydoc::Display)]
+pub enum StateStoreError {
+    /// couldn't create the parent directory
+    CreateDir(#[source] io::Error),
+    /// couldn't open the state file
+    Open(#[source] io::Error),
+    /// couldn't load state
+    Load(#[source] io::Error),
+    /// couldn't append entry to state
+    Append(#[source] io::Error),
+    /// couldn't append store the state
+    Store(#[source] io::Error),
+    /// couldn't load state
+    Serialize(#[source] serde_json::Error),
+    /// couldn't load state
+    Deserialize(#[source] serde_json::Error),
+}
+
+type Result<T> = std::result::Result<T, StateStoreError>;
+
+/// Handle to persist the state.
+///
+/// The file is a new line delimited JSON.
+#[derive(Debug)]
+pub struct StateStore {
+    file: BufWriter<File>,
+}
+
+impl StateStore {
+    /// Opens the file to use as store.
+    pub(crate) async fn open(file: impl AsRef<Path>) -> Result<Self> {
+        let path = file.as_ref();
+
+        if let Some(dir) = path.parent() {
+            tokio::fs::create_dir_all(dir)
+                .await
+                .map_err(StateStoreError::CreateDir)?;
+        }
+
+        let file = File::open(file).await.map_err(StateStoreError::Open)?;
+
+        Ok(Self {
+            file: BufWriter::new(file),
+        })
+    }
+
+    /// Load the state from the persistence
+    pub(crate) async fn load(&self) -> Result<()> {
+        // The call to read is one at the beginning, so we don't need to keep the reader around
+        let file = self
+            .file
+            .get_ref()
+            .try_clone()
+            .await
+            .map_err(StateStoreError::Load)?;
+
+        let mut reader = BufReader::new(file);
+        let mut line = String::new();
+
+        loop {
+            let byte_read = reader
+                .read_line(&mut line)
+                .await
+                .map_err(StateStoreError::Load)?;
+
+            if byte_read == 0 {
+                break;
+            }
+
+            let value: Value = serde_json::from_str(&line).map_err(StateStoreError::Deserialize)?;
```

Comment:
Should `value` be stored/loaded  to `self ` state?Maybe is leaked, or is it only loaded for checking file integration?

Reply:
It should be returned or be stored into the nodes struct, but I am not sure at the moment which direction to take. I was thinking of adding a todo here